### PR TITLE
Add capability of sending packets with size greater than 16#ffffff

### PR DESCRIPTION
**Pull request by skaee**
A little add-on marked as Todo in the code:
- Add capability of sending packets with size greater than 16#ffffff

### CODE REVIEW

File: src/mysql_protocol.erl [hunk 1]
```diff
@@ -1259,6 +1266,23 @@ add_packet_headers_test() ->
     {Data, 43} = add_packet_headers(<<"foo">>, 42),
     ?assertEqual(<<3, 0, 0, 42, "foo">>, list_to_binary(Data)).
 
+add_packet_headers_equal_to_0xffffff_test() ->
+	BigBin = erlang:list_to_binary([16#01 || I <- lists:seq(1, 16#ffffff)]),
+    {Data, 43} = add_packet_headers(BigBin, 42),
+    ?assertEqual(<<16#ff, 16#ff, 16#ff, 42, BigBin/binary>>, list_to_binary(Data)).
```
Comment (zuiderkwast) [inline src/mysql_protocol.erl]: This is not the correct behaviour. A packet size of 16#ffffff means that one more packet will come. There is no other indicator of if there are more packets to come or not. Thus, a packet of size 0 has to follow. See https://dev.mysql.com/doc/dev/mysql-server/latest/page_protocol_basic_packets.html .

In the implementation, it has to be `>=` in the guard, not `>`.

File: src/mysql_protocol.erl [hunk 2]
```diff
@@ -1018,11 +1018,18 @@ parse_packet_header(<<PacketLength:24/little-integer, SeqNum:8/integer>>) ->
...
-    %% Todo: implement the case when Size >= 16#ffffff.
-    if Size < 16#ffffff ->
-        {[<<Size:24/little, SeqNum:8>>, Bin], SeqNum1}
-    end.
+    add_packet_headers(Bin, Size, <<>>, SeqNum).
+
+%% @doc Manages all sizes of datas to send even with sizes greater than 0xffffff (16#ffffff).
+-spec add_packet_headers(PacketBody :: iodata(), Size :: integer(), Acc :: iodata(), SeqNum :: integer()) ->
+    {PacketWithHeaders :: iodata(), NextSeqNum :: integer()}.
+add_packet_headers(Bin, Size, Acc, SeqNum) when Size > 16#ffffff ->
```
Comment (zuiderkwast) [inline src/mysql_protocol.erl]: Why not use `byte_size(Bin)` in the guard? Then the Size parameter can be removed.

File: src/mysql_protocol.erl [hunk 3]
```diff
@@ -1018,11 +1018,18 @@ parse_packet_header(<<PacketLength:24/little-integer, SeqNum:8/integer>>) ->
...
     Bin = iolist_to_binary(PacketBody),
     Size = size(Bin),
-    SeqNum1 = (SeqNum + 1) band 16#ff,
-    %% Todo: implement the case when Size >= 16#ffffff.
-    if Size < 16#ffffff ->
-        {[<<Size:24/little, SeqNum:8>>, Bin], SeqNum1}
-    end.
+    add_packet_headers(Bin, Size, <<>>, SeqNum).
+
+%% @doc Manages all sizes of datas to send even with sizes greater than 0xffffff (16#ffffff).
```
Comment (zuiderkwast) [inline src/mysql_protocol.erl]: This comment looks like a commit message, something for the change log. For describing the function, something in line with "helper for add_packet_headers/2" would be more useful, I think.

File: src/mysql_protocol.erl [hunk 4]
```diff
@@ -1018,11 +1018,18 @@ parse_packet_header(<<PacketLength:24/little-integer, SeqNum:8/integer>>) ->
...
+-spec add_packet_headers(PacketBody :: iodata(), Size :: integer(), Acc :: iodata(), SeqNum :: integer()) ->
+    {PacketWithHeaders :: iodata(), NextSeqNum :: integer()}.
+add_packet_headers(Bin, Size, Acc, SeqNum) when Size > 16#ffffff ->
+    NextSeqNum = (SeqNum + 1) band 16#ff,
+    << Payload:16#ffffff/binary, Rest/binary >> = Bin,
+    add_packet_headers(Rest, Size - 16#ffffff, << Acc/binary, 16#ff, 16#ff, 16#ff, SeqNum:8, Payload/binary >>, NextSeqNum);
```
Comment (zuiderkwast) [inline src/mysql_protocol.erl]: The size bytes could be written as `16#ffffff:24/little`, i.e. in the same way as the packet size is coded in the other function clause.

This line is too long also. Please split into 80.

File: src/mysql_protocol.erl (hunk 2 above)
Comment (zuiderkwast) [reply src/mysql_protocol.erl]: I would prefer to have the `Acc` as the last parameter, just by convention.

File: src/mysql_protocol.erl (hunk 1 above)
Comment (zuiderkwast) [reply src/mysql_protocol.erl]: Nice test cases by the way. I like them.

Comment (krant) [reply src/mysql_protocol.erl]: We are also using custom patched version of mysql-otp purely because of limited packet size in upstream.

Comment (krant) [reply src/mysql_protocol.erl]: Nice lib by the way. I like it :)

Comment (zuiderkwast) [reply src/mysql_protocol.erl]: Good to know that more people/projects need this. I would like to merge soon. (Sorry for being picky!) :-)

**Comment by qingchuwudi** [comment]
Emysql's commit [fix: when a mysql **Query** is greater than 16m, it will be crashed.](https://github.com/qingchuwudi/Emysql/commit/e831e84dd8b162bc13cea0efca666c5e1a944b9e) maybe helpful.

**Comment by zuiderkwast** [comment]
Merged. Thanks! (I did some changes afterwards.)

**Pull request merged**